2020/4/passport/src/main.rs

```diff
@@ -0,0 +1,248 @@
+use std::env;
+use std::fs::File;
+use std::io::{self, BufRead};
+
+fn main() {
+    let args: Vec<String> = env::args().collect();
+    let file = File::open(&args[1]);
+
+    let reader = io::BufReader::new(file.unwrap());
+    let mut current = String::from("");
+    let mut passports = Vec::new();
+    for line in reader.lines() {
+        let l = line.unwrap();
+        if l.len() == 0 {
+            let passport = Passport::from_string(&current);
+            passports.push(passport);
+            current = String::from("");
+        } else {
+            current = current + " " + &l;
+        }
+    }
+    let passport = Passport::from_string(&current);
+    passports.push(passport);
+
+    println!("Found {} total passports", passports.len());
+    println!(
+        "Found {} valid passports",
+        passports.iter().filter(|p| p.is_valid()).count()
+    );
+}
+
+pub struct Passport {
+    birth_year: Option<String>,
+    issue_year: Option<String>,
+    expiration_year: Option<String>,
+    height: Option<String>,
+    hair_color: Option<String>,
+    eye_color: Option<String>,
+    passport_id: Option<String>,
+    country_id: Option<String>,
+}
+
+impl Passport {
+    pub fn from_string(data: &str) -> Passport {
+        let mut birth_year = None;
+        let mut issue_year = None;
+        let mut expiration_year = None;
+        let mut height = None;
+        let mut hair_color = None;
+        let mut eye_color = None;
+        let mut passport_id = None;
+        let mut country_id = None;
+        let parts = data.split_whitespace();
+        for p in parts {
+            let mut items = p.split(':');
+            let name = items.next().unwrap();
```

Comment:
Another case where I'd do something like:

```rust
let name = items.next()?;
```

And then have `from_string` return `Result<Passport, Error>`.

Check out the section [Propagating Errors](https://doc.rust-lang.org/book/ch09-02-recoverable-errors-with-result.html#propagating-errors) in the Rust book.